src/components/Sheet/Sheet.tsx

```diff
@@ -0,0 +1,180 @@
+  exitActive: classNames(styles.Bottom, styles.exitBottomActive),
+};
+
+export const RIGHT_CLASS_NAMES = {
+  enter: classNames(styles.Right, styles.enterRight),
+  enterActive: classNames(styles.Right, styles.enterRightActive),
+  exit: classNames(styles.Right, styles.exitRight),
+  exitActive: classNames(styles.Right, styles.exitRightActive),
+};
+
+export interface Props {
+  /** Whether or not the sheet is open */
+  open: boolean;
+  /** The child elements to render in the sheet */
+  children: React.ReactNode;
+  /** Callback when the backdrop is clicked or `ESC` is pressed */
+  onClose(): void;
+}
+
+export type ComposedProps = Props & WithAppProviderProps;
+
+export interface State {
+  mobile: boolean;
+}
+
+export class Sheet extends React.Component<ComposedProps, State> {
+  static contextTypes = frameContextTypes;
+  context: FrameContext;
+
+  state: State = {
+    mobile: false,
+  };
+
+  private handleResize = debounce(
+    () => {
+      const {
+        state: {mobile},
+        handleToggleMobile,
+      } = this;
+
+      if (mobile !== isMobile()) {
+        handleToggleMobile();
+      }
+    },
+    40,
+    {leading: true, trailing: true, maxWait: 40},
+  );
+
+  componentDidMount() {
+    const {
+      state: {mobile},
+      context: {frame},
+      props: {
+        polaris: {intl},
+      },
+      handleToggleMobile,
+    } = this;
+
+    if (frame == null) {
+      // eslint-disable-next-line no-console
+      console.warn(intl.translate('Polaris.Sheet.warningMessage'));
+    }
+
+    if (mobile !== isMobile()) {
+      handleToggleMobile();
+    }
+  }
+
+  render() {
+    const {
+      props: {children, open, onClose},
+      state: {mobile},
+      context: {frame},
+      handleResize,
+    } = this;
+
+    const sheetClassName = classNames(
+      styles.Sheet,
+      mobile === false && styles.desktop,
+    );
+
+    const containerClassName = classNames(
+      styles.Container,
+      mobile === false && styles.desktop,
+    );
+
+    const sharedTransitionProps = {
+      timeout: Duration.Slow,
+      in: open,
+      mountOnEnter: true,
+      unmountOnExit: true,
+    };
+
+    const finalTransitionProps = {
+      classNames: mobile ? BOTTOM_CLASS_NAMES : RIGHT_CLASS_NAMES,
+      ...sharedTransitionProps,
+    };
+
+    const backdropMarkup = open ? (
+      <Backdrop transparent onClick={onClose} />
+    ) : null;
+
+    if (frame == null) {
+      return null;
+    }
+
+    return (
+      <Portal idPrefix="sheet">
+        <CSSTransition {...finalTransitionProps}>
+          <Container
+            containerClassName={containerClassName}
+            open={open}
+            sheetClassName={sheetClassName}
+          >
+            {children}
+          </Container>
+        </CSSTransition>
+        <KeypressListener keyCode={Key.Escape} handler={onClose} />
+        <EventListener event="resize" handler={handleResize} />
+        {backdropMarkup}
```

Comment:
I wonder if it is worth it to abstract this–as someone scrolling through, it might make more sense to see the code from ln 124 down here, e.g.:

```{open && <Backdrop ...>```